File: src/2/6.cpp
#include <iostream>
#include "linked-list.cpp"

using namespace std;

// O(n) space & time
bool isPalindromeRecursive(Node *node, int count, Node **next)
{
    if (count == 0)
    {
        *next = node;
        return true;
    }

    if (node && count == 1)
    {
        *next = node->next;
        return true;
    }

    auto isMiddlePalindrome = isPalindromeRecursive(node->next, count - 2, next);
    if (!isMiddlePalindrome)
    {
        return false;
    }

    auto isPalindrome = (node->data == (*next)->data) && isMiddlePalindrome;
    *next = (*next)->next;
    return isPalindrome;
}

bool isPalindrome(LinkedList &list)
{
    Node *dummy = nullptr;
    return isPalindromeRecursive(list.head, list.length, &dummy);
}

int main()
{
    LinkedList test = {1, 2, 3, 3, 2, 1};
    cout << isPalindrome(test) << endl;

    LinkedList test1 = {1, 2, 3, 4, 5, 4, 3, 2, 1};
    cout << isPalindrome(test1) << endl;

    LinkedList test2 = {1, 2, 3, 4, 5, 5, 3, 2, 1};
    cout << isPalindrome(test2) << endl;

    LinkedList test3 = {99};
    cout << isPalindrome(test3) << endl;

    LinkedList test4 = {5, 7};
    cout << isPalindrome(test4) << endl;

    return 0;
}

File: src/2/6.js
const { LinkedList } = require("./linked-list.js");

const isPalindrome = (list) => {
  return isPalindromeRecursive(list.head, list.length);
};

const isPalindromeRecursive = (node, count, pair = { node: null }) => {
  if (count === 0) {
    pair.node = node;
    return true;
  }

  if (node && count === 1) {
    pair.node = node.next;
    return true;
  }

  const isMiddlePalindrome = isPalindromeRecursive(node.next, count - 2, pair);
  if (!isMiddlePalindrome) return false;
  const isPalindrome = node.data === pair.node.data && isMiddlePalindrome;
  pair.node = pair.node.next;

  return isPalindrome;
};

(() => {
  console.log(isPalindrome(new LinkedList([3, 6, 1, 1, 6, 3])));
  console.log(isPalindrome(new LinkedList([3, 6, 1, 7, 1, 6, 3])));
  console.log(isPalindrome(new LinkedList([3, 5])));
  console.log(isPalindrome(new LinkedList([3, 5, 2, 6, 9, 2, 5, 3])));
})();
